File: src/foam/nanos/auth/Authorizable.java
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.nanos.auth;

import foam.core.FObject;
import foam.core.X;

/**
 * A model should implement this interface if it is authorizable, meaning some
 * users are allowed to operate on (create, read, update, or delete) that object
 * but others are not.
 */
public interface Authorizable {
  void authorizeOnCreate(X x) throws AuthorizationException;
  void authorizeOnRead(X x) throws AuthorizationException;
  void authorizeOnUpdate(X x, FObject oldObj) throws AuthorizationException;
  void authorizeOnDelete(X x) throws AuthorizationException;
}

File: src/foam/java/Authorizable.js
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.INTERFACE({
  package: 'foam.nanos.auth',
  name: 'Authorizable',

  documentation: `
    A model should implement this interface if it is authorizable, meaning some
    users are allowed to operate on (create, read, update, or delete) that
    object but others are not.
  `,

  methods: [
    {
      name: 'authorizeOnCreate',
      args: [
        { name: 'x', type: 'Context' }
      ],
      javaThrows: ['foam.nanos.auth.AuthorizationException']
    },
    {
      name: 'authorizeOnRead',
      args: [
        { name: 'x', type: 'Context' }
      ],
      javaThrows: ['foam.nanos.auth.AuthorizationException']
    },
    {
      name: 'authorizeOnUpdate',
      args: [
        { name: 'x', type: 'Context' },
        { name: 'oldObj', type: 'foam.core.FObject' }
      ],
      javaThrows: ['foam.nanos.auth.AuthorizationException']
    },
    {
      name: 'authorizeOnDelete',
      args: [
        { name: 'x', type: 'Context' }
      ],
      javaThrows: ['foam.nanos.auth.AuthorizationException']
    }
  ]
});
